### PR TITLE
Delete outdated rendered master/worker machineconfigs and just keep the latest one

### PR DESCRIPTION
**Pull request by praveenkumar**
Those old rendered machine config contain the initial pull spec for
the cluster so better to remove those before bundle generation.

```
$ oc get mc --sort-by=.metadata.creationTimestamp --no-headers
99-master-ssh                                                                                 3.2.0   43h
chronyd-mask                                                                                  3.1.0   43h
99-worker-ssh                                                                                 3.2.0   43h
99-openshift-machineconfig-master-dummy-networks                                              3.2.0   43h
00-master                                          b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
00-worker                                          b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
01-master-kubelet                                  b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
01-master-container-runtime                        b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
01-worker-container-runtime                        b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
01-worker-kubelet                                  b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
rendered-master-6a52c924beeab10c4d917eb7786d8d9c   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
rendered-worker-c0d4ccd0599554ca84cdac121c99b8c0   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
99-worker-generated-registries                     b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
99-master-generated-registries                     b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   43h
rendered-worker-73ad21005c178f63a66398638aaa05ae   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   18h
rendered-master-35111bcd08dccd57f47acaa57cb1d296   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   18h
rendered-master-f5f05aad7d6a34e21fca04e799e54a44   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   3h21m
rendered-worker-24d71aa1fbe25a89567669d89d53065d   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   3h21m
rendered-master-de3a028583cb2ac28121fc8d527637a2   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   142m
rendered-worker-68ca5066a7c81354e3693a0abf8ff3df   b20ecdcd81dc3aa1269757ddfe97603918194b3b   3.2.0   142m

// All the rendered master except lastet one
$ oc get mc --sort-by=.metadata.creationTimestamp --no-headers -oname | grep rendered-master | head -n -1
machineconfig.machineconfiguration.openshift.io/rendered-master-6a52c924beeab10c4d917eb7786d8d9c
machineconfig.machineconfiguration.openshift.io/rendered-master-35111bcd08dccd57f47acaa57cb1d296
machineconfig.machineconfiguration.openshift.io/rendered-master-f5f05aad7d6a34e21fca04e799e54a44
```

**Fixes:** Issue #N

**Relates to:** Issue #N, PR #N, ...

## Solution/Idea

Describe in plain English what you solved and how. For instance, _Added `start` command to CRC so the user can create a VM and set-up a single-node OpenShift cluster on it with one command. It requires blablabla..._

## Proposed changes

List main as well as consequential changes you introduced or had to introduce.

1. Add `start` command.
2. Add `setup` as prerequisite to `start`.
3. ...

## Testing

What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded
2. stdout contains ... if `start` succeeded
3. stderr contains ... after `start` failed
4. `status` returns ... if `start` succeeded
5. `status` returns ... if `start` failed
6. `start` fails after `start` succeeded or after `status` says CRC is `Running`
7. ...

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 SHELL := /bin/bash
 
-BUNDLE_VERSION ?= 4.7.13
+BUNDLE_VERSION ?= 4.8.0-rc.0
```
Comment (praveenkumar) [inline Makefile:3]: This version update is only to test if current master along with this PR is able to provide us a green CI or do we still need some more changes to incorporate 4.8 release with crc.

File: pkg/crc/cluster/cluster.go [hunk 2]
```diff
@@ -168,18 +168,31 @@ func RemovePullSecretFromCluster(ocConfig oc.Config, sshRunner *ssh.Runner) erro
...
+	// in 4.7 we disabled mco and also deleted all the machineconfig/machineconfig-pools.
+	// For 4.8 we don't disable mco and it does contain the machineconfigs.
+	cmd := "oc --kubeconfig=/opt/kubeconfig get mc --sort-by=.metadata.creationTimestamp --no-headers -oname | grep rendered-master"
+	_, _, err := runner.Run(cmd)
+	if err != nil {
+		return nil
+	}
+
+	cmd = "oc --kubeconfig=/opt/kubeconfig get mc --sort-by=.metadata.creationTimestamp --no-headers -oname | grep rendered-master | head -n -1 | xargs -t oc --kubeconfig=/opt/kubeconfig delete"
+	_, stderr, err := runner.Run(cmd)
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:186]: Could these various `oc` commandlines make use of `OcRunner`/`OcConfig`?

Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go:186]: @cfergeau `OcRunner` doesn't support pipe so if we really go with `OcRunner` then we need to do something like following.
```
mc, stderr, err := ocConfig.RunOcCommand("get mc --sort-by=.metadata.creationTimestamp --no-headers -oname")
[...]
oldmc, stderr, err := sshRunner.Run(fmt.Sprintf("%s | grep rendered-master | head -n -1", mc)
[...]
_, stderr,err := ocConfig.RunOcCommand(fmt.Sprintf("delete %s", oldmc))
```

Comment (cfergeau) [reply pkg/crc/cluster/cluster.go:186]: We don't have to use `| grep| head` to achieve this, we could ask for json output and let go parse it (do we already vendor the module with the type definitions?), or use https://golang.org/pkg/bufio/#NewScanner to get what we are interested in?

Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go:186]: @cfergeau I will check again with json output.

**Comment by praveenkumar** [comment]
CI failure on `e2e-crc` because of https://bugzilla.redhat.com/show_bug.cgi?id=1979550 one.

File: pkg/crc/machine/generate_bundle.go [hunk 3]
```diff
@@ -33,17 +32,21 @@ func (client *client) GenerateBundle(forceStop bool) error {
...
+		} else {
+			return err
 		}
-		return err
 	}
-	if currentState != state.Stopped {
-		return fmt.Errorf("VM is not stopped, current state is %s", currentState.String())
+	running, err := client.IsRunning()
+	if err != nil {
+		return nil
```
Comment (cfergeau) [inline pkg/crc/machine/generate_bundle.go:46]: `return err`?

File: pkg/crc/machine/generate_bundle.go [hunk 4]
```diff
@@ -33,17 +32,21 @@ func (client *client) GenerateBundle(forceStop bool) error {
...
-	if err != nil {
+	if _, err := client.Stop(); err != nil {
 		if forceStop {
 			if err := client.PowerOff(); err != nil {
 				return err
 			}
+		} else {
+			return err
 		}
-		return err
```
Comment (cfergeau) [inline pkg/crc/machine/generate_bundle.go:39]: The commit logs says ` Fix the state of VM when generating the bundle ` but it seems this is also fixing  the fact that we'd always error out when forced shutdown was triggered, even if it was successful?

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,6 @@
 SHELL := /bin/bash
 
-BUNDLE_VERSION ?= 4.7.18
+BUNDLE_VERSION ?= 4.8.0
```
Comment (cfergeau) [inline Makefile:3]: no `-rcX` in the versioning?

File: pkg/crc/cluster/cluster.go [hunk 6]
```diff
@@ -192,18 +192,50 @@ func RemovePullSecretFromCluster(ocConfig oc.Config, sshRunner *ssh.Runner) erro
...
+	if len(sortedMachineConfigsWithTime) == 0 {
+		return nil
 	}
 
-	cmdArgs = []string{"delete", "machineconfigpools", "--all"}
-	_, stderr, err = ocConfig.RunOcCommand(cmdArgs...)
-	if err != nil {
-		return fmt.Errorf("Failed to remove machineconfigpools %w: %s", err, stderr)
+	var (
+		renderedMaster []string
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:213]: could be `renderedMasters` and `renderedWorkers` since an array can hold multiple masters/workers.

File: pkg/crc/cluster/cluster.go [hunk 7]
```diff
@@ -192,18 +192,50 @@ func RemovePullSecretFromCluster(ocConfig oc.Config, sshRunner *ssh.Runner) erro
...
+		deleteRenderedMachineConfig = fmt.Sprintf("%s %s", deleteRenderedMachineConfig, strings.Join(renderedWorker[:len(renderedWorker)-1], " "))
 	}
 
+	if deleteRenderedMachineConfig != "" {
+		_, stderr, err = ocConfig.RunOcCommand(fmt.Sprintf("delete %s", deleteRenderedMachineConfig))
+		if err != nil {
+			return fmt.Errorf("Failed to remove machineconfigpools %w: %s", err, stderr)
+		}
+	}
 	return nil
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:239]: I have only build-tested this, but I think the code below does the same while being simpler (less code overall):
```
func RemoveOldRenderedMachineConfig(ocConfig oc.Config) error {
        // This block (#L179-183) should be removed as soon as we start shipping with 4.8 bundle.
        // This check is only make sure if there is any machineconfig resource or not because
        // in 4.7 we disabled mco and also deleted all the machineconfig/machineconfig-pools.
        // For 4.8 we don't disable mco and it does contain the machineconfigs.    
        stdout, stderr, err := ocConfig.RunOcCommand("get mc --no-headers -oname")
        if err != nil {
                return fmt.Errorf("failed to get machineconfig resource %w: %s", err, stderr)
        }
        machineConfigs := strings.Split(stdout, "\n")
        
        var ocDeleteArgs = []string{"delete"}
                                                            
        for _, mc := range machineConfigs {                 
                if strings.Contains(mc, "rendered-master") {   
                        ocDeleteArgs = append(ocDeleteArgs, mc)
                }                                              
                if strings.Contains(mc, "rendered-worker") {       
                        ocDeleteArgs = append(ocDeleteArgs, mc)
                }
        }
        if len(ocDeleteArgs) == 1 {
                // no machine configs to delete
                return nil
        }
                
        _, stderr, err = ocConfig.RunOcCommand(ocDeleteArgs...)
        if err != nil {
                return fmt.Errorf("Failed to remove machineconfigpools %w: %s", err, stderr)
        }
        return nil
}
```

Comment (cfergeau) [reply pkg/crc/cluster/cluster.go:239]: (`Contains` could probably be `StartsWith`/`HasSuffix`?

**Review by cfergeau — COMMENTED** [review]
Few more comments, but overall looking good :+1:

File: pkg/crc/cluster/cluster.go (hunk 7 above)
Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go:240]: @cfergeau the above code will delete all the mc which have `rendered master` and `redered worker` but what we want only to delete the older one not the latest one.

File: pkg/crc/cluster/cluster.go (hunk 6 above)
Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go:214]: In our case we have only single master/worker, config may be multiple.

File: Makefile (hunk 5 above)
Comment (praveenkumar) [reply Makefile:3]: @cfergeau we already have 4.8.0 bundle.

File: pkg/crc/machine/generate_bundle.go (hunk 4 above)
Comment (praveenkumar) [reply pkg/crc/machine/generate_bundle.go:39]: @cfergeau yes fixed it, created new commit to handle it.

**Review by guillaumerose — APPROVED** [review]
